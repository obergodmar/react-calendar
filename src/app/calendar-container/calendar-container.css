.calendar {
    position: relative;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: space-between;
    box-sizing: border-box;
    width: 224px;
    height: 274px;
    padding: 13px;
    font-size: 12px;
    background-color: #fff;
    border: 1px solid gray;
    box-shadow: 0 3px 6px rgba(0, 0, 0, .16), 0 3px 6px rgba(0, 0, 0, .23);
    user-select: none;
}

.todayButton {
    align-self: flex-end;
    margin-top: 5px;
    margin-right: 7px;
    color: #0078d4;
    font-size: 13px;
    background-color: #fff;
    border: none;
}

.todayButton:hover {
    background-color: #b2d5ff;
}
