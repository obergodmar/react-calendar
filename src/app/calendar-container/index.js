export { CalendarContainer } from "./calendar-container";
