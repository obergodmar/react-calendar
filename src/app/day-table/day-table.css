.row {
    display: flex;
    justify-content: space-between;
    color: black;
    text-align: center;
}

.list,
.row {
    list-style: none;
}

.weekday {
    width: 28px;
    height: 20px;
}

.day {
    position: relative;
    z-index: 1;
    display: flex;
    align-items: center;
    justify-content: center;
    box-sizing: border-box;
    width: 28px;
    height: 28px;
    border: 1px solid transparent;
    border-radius: 2px;
    cursor: default;
}

.day:focus {
    outline: none;
}

.day:focus,
.day:hover {
    background-color: #b2d5ff;
}

.dayOther {
    color: gray;
}

.dayToday {
    border: 1px solid black;
}

.daySelected {
    color: #fff;
    font-weight: 600;
    background-color: #0075ff;
    border: 1px solid black;
    border-radius: 0;
}

.daySelected:focus,
.daySelected:hover {
    background-color: #0075ff;
    border-radius: 0;
}
