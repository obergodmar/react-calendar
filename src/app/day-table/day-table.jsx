import React, { Fragment, useCallback, useMemo } from "react";
import PropTypes from "prop-types";
import cn from "classnames";
import format from "date-fns/format";
import addDays from "date-fns/addDays";
import startOfWeek from "date-fns/startOfWeek";

import { TEST_DATE, WEEKDAY_FORMAT } from "../../utils";

import style from "./day-table.css";

export const DayTable = ({ calendar, changeDate }) => {
  // TODO: Add locale
  const weekdays = useMemo(() => {
    let date = startOfWeek(TEST_DATE);
    return [...Array(7).keys()].map(() => {
      const weekday = format(date, WEEKDAY_FORMAT);
      date = addDays(date, 1);
      return weekday;
    });
  }, []);

  const handlePreventDefault = useCallback(
    (event) => event.preventDefault(),
    []
  );

  const handleKeyDown = useCallback(
    (event, date, needClose) => {
      const { keyCode } = event;
      if (keyCode === 13 || keyCode === 32) {
        changeDate(date, needClose);
      }
    },
    [changeDate]
  );

  return (
    <Fragment>
      <ul className={style.row}>
        {weekdays.map((day, idx) => (
          <li key={idx} className={style.weekday}>
            {day}
          </li>
        ))}
      </ul>
      <ul className={style.list}>
        {calendar.map((row, rowIdx) => (
          <li key={rowIdx}>
            <ul className={style.row}>
              {row.map((day, dayIdx) => (
                <li
                  tabIndex={0}
                  onMouseDown={handlePreventDefault}
                  onClick={() => changeDate(day.date, day.isPart)}
                  onKeyDown={(e) => handleKeyDown(e, day.date, day.isPart)}
                  key={dayIdx}
                  className={cn(style.day, {
                    [style.dayOther]: !day.isPart,
                    [style.dayToday]: day.isToday,
                    [style.daySelected]: day.isSelected,
                  })}
                >
                  {day.date.getDate()}
                </li>
              ))}
            </ul>
          </li>
        ))}
      </ul>
    </Fragment>
  );
};

DayTable.propTypes = {
  calendar: PropTypes.arrayOf(
    PropTypes.arrayOf(
      PropTypes.shape({
        date: PropTypes.instanceOf(Date),
        isToday: PropTypes.bool,
        isSelected: PropTypes.bool,
        isPart: PropTypes.bool,
      })
    )
  ).isRequired,
  changeDate: PropTypes.func.isRequired,
};

DayTable.displayName = "DayTable";
