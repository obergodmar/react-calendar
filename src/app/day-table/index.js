export { DayTable } from "./day-table";
