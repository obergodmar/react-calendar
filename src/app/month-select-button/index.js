export { MonthSelectButton } from "./month-select-button";
