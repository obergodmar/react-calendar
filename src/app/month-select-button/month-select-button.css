.buttonGroup {
    display: flex;
    align-items: center;
    align-self: flex-start;
    justify-content: space-between;
    width: 100%;
    height: 20px;
    margin-bottom: 13px;
}

.buttonUpDown {
    display: flex;
    align-items: center;
}

.buttonUp,
.buttonDown {
    width: 25px;
    height: 20px;
    font-family: Roboto, sans-serif;
    background:
        url("data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAMAAABEpIrGAAAACXBIWXMAAA8AAAAPAAEXcSFwAAAAGXRFWHRTb2Z0d2FyZQB3d3cuaW5rc2NhcGUub3Jnm+48GgAAAGNQTFRFAAAAKysrJCQkJycUJCQkICAgHR0dHBwcHh4eHR0dHh4eHh4eHSMdHCEcHh4eHiIeHSEdHSAdHx8cHx8cHiIeHSAdHiAeHiAcHSAdHiAdHyAdHiAdHiAdHiAdHiAdHiAdHiAdXqmzOwAAACB0Uk5TAAYHDQ4YGhsiIyorLC5DREZHSUpMaW/O0NbZ2vf4+f21tk5jAAAAiElEQVQ4y+3QWxaCIBSF4W0WaXfNkkRyz3+UaUki68gIPIsn/o/zADAbpbVCZHYv0uzjPSaUYcf+GLX4/v0kH+3CjqFXOZnfZPHtGABE0Xdb4QckMfYRoAiF6w6EYtuwLeADlJZN+gdH1yeAu+VhWnE9IQQ4X6TP8IA8K1iBPxmZRcFG18n85gPxehI6iMeA4gAAAABJRU5ErkJggg==")
        no-repeat 50%,
        #fff;
    background-size: 15px;
    border: none;
}

.buttonUp:hover,
.buttonDown:hover {
    background-color: #b2d5ff;
}

.buttonDown {
    margin-left: 5px;
    transform: rotate(180deg);
}

.monthSelectButton {
    display: flex;
    align-items: center;
    padding: 0;
    color: black;
    font-weight: 600;
    line-height: 1.53;
    background-color: #fff;
    border: none;
    border-radius: 2px;
}

.selectIcon {
    width: 0;
    height: 0;
    margin-left: 5px;
    border-top: 4px solid #2c3136;
    border-right: 4px solid transparent;
    border-left: 4px solid transparent;
}

.selectShown {
    color: lightgray;
}

.selectIconShown {
    border-top-color: lightgray;
}
