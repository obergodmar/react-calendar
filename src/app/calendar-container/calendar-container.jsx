import React, { Fragment, useCallback, useMemo, useState } from "react";
import PropTypes from "prop-types";
import set from "date-fns/set";
import format from "date-fns/format";
import addMonths from "date-fns/addMonths";
import addYears from "date-fns/addYears";

import { dayMatrix, MONTH_YEAR } from "../../utils";
import { MonthSelect } from "../month-select";
import { MonthSelectButton } from "../month-select-button";
import { YearPicker } from "../year-picker";
import { DayTable } from "../day-table";

import style from "./calendar-container.css";

export const CalendarContainer = ({ date, changeDate }) => {
  const [isMonthSelectShown, setMonthSelectShown] = useState(false);
  const calendar = useMemo(() => dayMatrix(date), [date]);
  const currentDate = useMemo(() => {
    const monthYear = format(date, MONTH_YEAR);
    return monthYear.charAt(0).toUpperCase() + monthYear.slice(1);
  }, [date]);

  const handleChangeDate = useCallback(
    (newDate, needClose = false) => {
      changeDate(newDate, needClose);
    },
    [changeDate]
  );

  const handleShowMonthSelect = useCallback(() => {
    setMonthSelectShown(!isMonthSelectShown);
  }, [isMonthSelectShown]);

  const handleChangeMonth = useCallback(
    (month) => {
      const newDate = set(date, { month });
      handleChangeDate(newDate);
      handleShowMonthSelect();
    },
    [date, handleChangeDate, handleShowMonthSelect]
  );

  const handleAddMonth = useCallback(
    (amount) => {
      const newDate = addMonths(date, amount);
      handleChangeDate(newDate);
    },
    [date, handleChangeDate]
  );

  const handleAddYear = useCallback(
    (amount) => {
      const newDate = addYears(date, amount);
      handleChangeDate(newDate);
    },
    [date, handleChangeDate]
  );

  const handleChangeToToday = useCallback(() => {
    handleChangeDate(new Date(), true);
  }, [handleChangeDate]);

  return (
    <div className={style.calendar}>
      <MonthSelectButton
        isShown={isMonthSelectShown}
        addMonth={handleAddMonth}
        onClick={handleShowMonthSelect}
        currentDate={currentDate}
      />
      {isMonthSelectShown ? (
        <Fragment>
          <YearPicker date={date.getFullYear()} addYear={handleAddYear} />
          <MonthSelect changeMonth={handleChangeMonth} date={date} />
        </Fragment>
      ) : (
        <Fragment>
          <DayTable calendar={calendar} changeDate={handleChangeDate} />
          <button onClick={handleChangeToToday} className={style.todayButton}>
            Today
          </button>
        </Fragment>
      )}
    </div>
  );
};

CalendarContainer.propTypes = {
  date: PropTypes.instanceOf(Date).isRequired,
  changeDate: PropTypes.func.isRequired,
};

CalendarContainer.displayName = "CalendarContainer";
