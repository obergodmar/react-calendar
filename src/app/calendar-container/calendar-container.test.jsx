import React from "react";
import { shallow } from "enzyme";
import { describe, jest, test } from "@jest/globals";
import { YearPicker } from "../year-picker";
import { MonthSelect } from "../month-select";
import { DayTable } from "../day-table";
import { MonthSelectButton } from "../month-select-button";

import { CalendarContainer } from "./calendar-container";
import { TEST_DATE } from "../../utils";

describe("Calendar app", () => {
  const handleChangeDate = jest.fn();
  const date = new Date(TEST_DATE);

  const wrapper = shallow(
    <CalendarContainer changeDate={handleChangeDate} date={date} />
  );

  test("Should display DayTable by default", () => {
    expect(wrapper.find(YearPicker)).toHaveLength(0);
    expect(wrapper.find(MonthSelect)).toHaveLength(0);
    expect(wrapper.find(DayTable)).toHaveLength(1);
    expect(wrapper.find(MonthSelectButton)).toHaveLength(1);
    expect(wrapper).toMatchSnapshot();
  });

  test("Should display YearPicker and MonthSelect when click on button", () => {
    wrapper.find(MonthSelectButton).props().onClick();
    expect(wrapper.find(YearPicker)).toHaveLength(1);
    expect(wrapper.find(MonthSelect)).toHaveLength(1);
    expect(wrapper.find(DayTable)).toHaveLength(0);
    expect(wrapper.find(MonthSelectButton)).toHaveLength(1);
    expect(wrapper).toMatchSnapshot();
  });

  test("Should change date", () => {
    const mockChangeDate = jest.fn();
    const component = shallow(
      <CalendarContainer changeDate={mockChangeDate} date={date} />
    );

    component.find(MonthSelectButton).props().addMonth();
    expect(mockChangeDate).toHaveBeenCalledTimes(1);

    component.find(DayTable).props().changeDate();
    expect(mockChangeDate).toHaveBeenCalledTimes(2);

    component.find("button").simulate("click");
    expect(mockChangeDate).toHaveBeenCalledTimes(3);

    component.find(MonthSelectButton).props().onClick();
    component.find(YearPicker).props().addYear();
    expect(mockChangeDate).toHaveBeenCalledTimes(4);

    component.find(MonthSelect).props().changeMonth();
    expect(mockChangeDate).toHaveBeenCalledTimes(5);
  });
});
