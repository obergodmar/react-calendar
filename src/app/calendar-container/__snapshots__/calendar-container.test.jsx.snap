// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar app Should display DayTable by default 1`] = `
<div
  className="calendar"
>
  <MonthSelectButton
    addMonth={[Function]}
    currentDate="February 1970"
    isShown={false}
    onClick={[Function]}
  />
  <DayTable
    calendar={
      Array [
        Array [
          Object {
            "date": 1970-01-31T21:00:00.000Z,
            "isPart": true,
            "isSelected": true,
            "isToday": false,
          },
          Object {
            "date": 1970-02-01T21:00:00.000Z,
            "isPart": true,
            "isSelected": false,
            "isToday": false,
          },
          Object {
            "date": 1970-02-02T21:00:00.000Z,
            "isPart": true,
            "isSelected": false,
            "isToday": false,
          },
          Object {
            "date": 1970-02-03T21:00:00.000Z,
            "isPart": true,
            "isSelected": false,
            "isToday": false,
          },
          Object {
            "date": 1970-02-04T21:00:00.000Z,
            "isPart": true,
            "isSelected": false,
            "isToday": false,
          },
          Object {
            "date": 1970-02-05T21:00:00.000Z,
            "isPart": true,
            "isSelected": false,
            "isToday": false,
          },
          Object {
            "date": 1970-02-06T21:00:00.000Z,
            "isPart": true,
            "isSelected": false,
            "isToday": false,
          },
        ],
        Array [
          Object {
            "date": 1970-02-07T21:00:00.000Z,
            "isPart": true,
            "isSelected": false,
            "isToday": false,
          },
          Object {
            "date": 1970-02-08T21:00:00.000Z,
            "isPart": true,
            "isSelected": false,
            "isToday": false,
          },
          Object {
            "date": 1970-02-09T21:00:00.000Z,
            "isPart": true,
            "isSelected": false,
            "isToday": false,
          },
          Object {
            "date": 1970-02-10T21:00:00.000Z,
            "isPart": true,
            "isSelected": false,
            "isToday": false,
          },
          Object {
            "date": 1970-02-11T21:00:00.000Z,
            "isPart": true,
            "isSelected": false,
            "isToday": false,
          },
          Object {
            "date": 1970-02-12T21:00:00.000Z,
            "isPart": true,
            "isSelected": false,
            "isToday": false,
          },
          Object {
            "date": 1970-02-13T21:00:00.000Z,
            "isPart": true,
            "isSelected": false,
            "isToday": false,
          },
        ],
        Array [
          Object {
            "date": 1970-02-14T21:00:00.000Z,
            "isPart": true,
            "isSelected": false,
            "isToday": false,
          },
          Object {
            "date": 1970-02-15T21:00:00.000Z,
            "isPart": true,
            "isSelected": false,
            "isToday": false,
          },
          Object {
            "date": 1970-02-16T21:00:00.000Z,
            "isPart": true,
            "isSelected": false,
            "isToday": false,
          },
          Object {
            "date": 1970-02-17T21:00:00.000Z,
            "isPart": true,
            "isSelected": false,
            "isToday": false,
          },
          Object {
            "date": 1970-02-18T21:00:00.000Z,
            "isPart": true,
            "isSelected": false,
            "isToday": false,
          },
          Object {
            "date": 1970-02-19T21:00:00.000Z,
            "isPart": true,
            "isSelected": false,
            "isToday": false,
          },
          Object {
            "date": 1970-02-20T21:00:00.000Z,
            "isPart": true,
            "isSelected": false,
            "isToday": false,
          },
        ],
        Array [
          Object {
            "date": 1970-02-21T21:00:00.000Z,
            "isPart": true,
            "isSelected": false,
            "isToday": false,
          },
          Object {
            "date": 1970-02-22T21:00:00.000Z,
            "isPart": true,
            "isSelected": false,
            "isToday": false,
          },
          Object {
            "date": 1970-02-23T21:00:00.000Z,
            "isPart": true,
            "isSelected": false,
            "isToday": false,
          },
          Object {
            "date": 1970-02-24T21:00:00.000Z,
            "isPart": true,
            "isSelected": false,
            "isToday": false,
          },
          Object {
            "date": 1970-02-25T21:00:00.000Z,
            "isPart": true,
            "isSelected": false,
            "isToday": false,
          },
          Object {
            "date": 1970-02-26T21:00:00.000Z,
            "isPart": true,
            "isSelected": false,
            "isToday": false,
          },
          Object {
            "date": 1970-02-27T21:00:00.000Z,
            "isPart": true,
            "isSelected": false,
            "isToday": false,
          },
        ],
        Array [
          Object {
            "date": 1970-02-28T21:00:00.000Z,
            "isPart": false,
            "isSelected": false,
            "isToday": false,
          },
          Object {
            "date": 1970-03-01T21:00:00.000Z,
            "isPart": false,
            "isSelected": false,
            "isToday": false,
          },
          Object {
            "date": 1970-03-02T21:00:00.000Z,
            "isPart": false,
            "isSelected": false,
            "isToday": false,
          },
          Object {
            "date": 1970-03-03T21:00:00.000Z,
            "isPart": false,
            "isSelected": false,
            "isToday": false,
          },
          Object {
            "date": 1970-03-04T21:00:00.000Z,
            "isPart": false,
            "isSelected": false,
            "isToday": false,
          },
          Object {
            "date": 1970-03-05T21:00:00.000Z,
            "isPart": false,
            "isSelected": false,
            "isToday": false,
          },
          Object {
            "date": 1970-03-06T21:00:00.000Z,
            "isPart": false,
            "isSelected": false,
            "isToday": false,
          },
        ],
        Array [
          Object {
            "date": 1970-03-07T21:00:00.000Z,
            "isPart": false,
            "isSelected": false,
            "isToday": false,
          },
          Object {
            "date": 1970-03-08T21:00:00.000Z,
            "isPart": false,
            "isSelected": false,
            "isToday": false,
          },
          Object {
            "date": 1970-03-09T21:00:00.000Z,
            "isPart": false,
            "isSelected": false,
            "isToday": false,
          },
          Object {
            "date": 1970-03-10T21:00:00.000Z,
            "isPart": false,
            "isSelected": false,
            "isToday": false,
          },
          Object {
            "date": 1970-03-11T21:00:00.000Z,
            "isPart": false,
            "isSelected": false,
            "isToday": false,
          },
          Object {
            "date": 1970-03-12T21:00:00.000Z,
            "isPart": false,
            "isSelected": false,
            "isToday": false,
          },
          Object {
            "date": 1970-03-13T21:00:00.000Z,
            "isPart": false,
            "isSelected": false,
            "isToday": false,
          },
        ],
      ]
    }
    changeDate={[Function]}
  />
  <button
    className="todayButton"
    onClick={[Function]}
  >
    Today
  </button>
</div>
`;

exports[`Calendar app Should display YearPicker and MonthSelect when click on button 1`] = `
<div
  className="calendar"
>
  <MonthSelectButton
    addMonth={[Function]}
    currentDate="February 1970"
    isShown={true}
    onClick={[Function]}
  />
  <YearPicker
    addYear={[Function]}
    date={1970}
  />
  <MonthSelect
    changeMonth={[Function]}
    date={1970-01-31T21:00:00.000Z}
  />
</div>
`;
